File: lab10/src/by/bntu/fitr/minigame/domain/east/DarkElf.java
package by.bntu.fitr.minigame.domain.east;

import by.bntu.fitr.minigame.domain.Creature;
import by.bntu.fitr.minigame.domain.Role;
import by.bntu.fitr.minigame.domain.Weapon;

public class DarkElf extends Creature {
    private Weapon weapon = Weapon.STAFF;
    private Role role = Role.SORCERER;

    public DarkElf(int health, int strength, int agility, int intelligence) {
        super(health, strength, agility, intelligence);
    }

    @Override
    protected int getWeaponDamage() {
        return weapon.getDamage();
    }

    @Override
    public int buff() {
        return (role.getPercent() + weapon.getDamage()) / 100 * getIntelligence();
    }
}

File: lab10kt/src/main/kotlin/by/bntu/fitr/povt/korotkevich/model/east/DarkElf.kt
package by.bntu.fitr.povt.korotkevich.model.east

import by.bntu.fitr.povt.korotkevich.model.Creature
import by.bntu.fitr.povt.korotkevich.model.Role
import by.bntu.fitr.povt.korotkevich.model.Weapon
import by.bntu.fitr.povt.korotkevich.model.buff

class DarkElf(override val health: Int = 80, override val strength: Int = 5,
              override val agility: Int = 5, override val intelligence: Int = 50,
              val role: Role = Role.SORCERER, val weapon: Weapon = Weapon.STUFF) : Creature() {

    override val id: String = "DarkElf"

    override fun getSelfDamage() = weapon.damage + buff(role)

    override fun toString() = StringBuilder("DarkElf(").apply {
        append("${super.toString()}, role: $role, weapon: $weapon)")
    }.toString()
}